File: Gagan_Namdev_Capstone_Issue_Sprint_Management_System/backend/app/services/project_service.py
from app.models.project_model import ProjectModel
from app.repositories.project_repository import ProjectRepository


class ProjectService:

    @staticmethod
    def create_project(project):

        existing_project = ProjectRepository.get_project_by_name(
            project.name
        )

        if existing_project:
            return {
                "message": "Project already exists"
            }

        project_data = ProjectModel.create_project(
            name=project.name,
            description=project.description,
            members=project.members
        )

        ProjectRepository.create_project(project_data)

        return {
            "message": "Project created successfully",
            "project_key": project_data["project_key"]
        }

    @staticmethod
    def get_all_projects():

        projects = ProjectRepository.get_all_projects()

        result = []

        for project in projects:

            result.append({
                "_id": str(project["_id"]),
                "project_key": project.get("project_key"),
                "name": project.get("name"),
                "description": project.get("description"),
                "members": project.get("members", []),
                "created_at": project.get("created_at")
            })

        return result

    @staticmethod
    def get_my_projects(current_user):

        projects = ProjectRepository.get_all_projects()

        result = []

        for project in projects:

            if current_user["email"] in project.get("members", []):

                result.append({
                    "_id": str(project["_id"]),
                    "project_key": project.get("project_key"),
                    "name": project.get("name"),
                    "description": project.get("description"),
                    "members": project.get("members", []),
                    "created_at": project.get("created_at")
                })

        return result

    @staticmethod
    def get_project_by_id(project_id: str):

        project = ProjectRepository.get_project_by_id(
            project_id
        )

        if not project:
            return {
                "message": "Project not found"
            }

        return {
            "_id": str(project["_id"]),
            "project_key": project.get("project_key"),
            "name": project.get("name"),
            "description": project.get("description"),
            "members": project.get("members", []),
            "created_at": project.get("created_at")
        }

    @staticmethod
    def update_project(project_id: str, project):

        existing_project = ProjectRepository.get_project_by_id(
            project_id
        )

        if not existing_project:
            return {
                "message": "Project not found"
            }

        ProjectRepository.update_project(
            project_id,
            {
                "name": project.name,
                "description": project.description,
                "members": project.members
            }
        )

        return {
            "message": "Project updated successfully"
        }

    @staticmethod
    def delete_project(project_id: str):

        existing_project = ProjectRepository.get_project_by_id(
            project_id
        )

        if not existing_project:
            return {
                "message": "Project not found"
            }

        ProjectRepository.delete_project(project_id)

        return {
            "message": "Project deleted successfully"
        }

    @staticmethod
    def assign_members(project_id: str, data):

        project = ProjectRepository.get_project_by_id(
            project_id
        )

        if not project:
            return {
                "message": "Project not found"
            }

        ProjectRepository.assign_members(
            project_id,
            data.members
        )

        return {
            "message": "Members assigned successfully"
        }

File: Gagan_Namdev_Capstone_Issue_Sprint_Management_System/frontend/src/services/project-service.js
import api from "../config/axios";

export const createProject = async (data) => {

    const response = await api.post(
        "/projects/",
        data
    );

    return response.data;

};

export const getProjects = async () => {

    const response = await api.get(
        "/projects/"
    );

    return response.data;

};

export const getMyProjects = async () => {

    const response = await api.get(
        "/projects/my"
    );

    return response.data;

};

export const getProjectById = async (id) => {

    const response = await api.get(
        `/projects/${id}`
    );

    return response.data;

};

export const updateProject = async (id, data) => {

    const response = await api.put(

        `/projects/${id}`,

        data

    );

    return response.data;

};

export const deleteProject = async (id) => {

    const response = await api.delete(

        `/projects/${id}`

    );

    return response.data;

};

export const assignMembers = async (

    id,

    members

) => {

    const response = await api.put(

        `/projects/${id}/members`,

        {

            members,

        }

    );

    return response.data;

};